### PR TITLE
Find x by y / m.cc security

### LINKED ISSUE #315: Gate reusing of existing roles and channels by m.cc

### LINKED ISSUE #332: Reduce unecessary duplication of DB find commands

### PR DESCRIPTION
**Pull request by cimspin**
Fixes #315 
Resolves #332

### CODE REVIEW

File: MissionAlertBot.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 import ast
 import copy
 from doctest import debug_script
+from msilib import knownbits
```
Comment (gcruicks) [inline MissionAlertBot.py:10]: Not sure what you are trying to do here, you don't seem to use `knownbits` anywhere and msilib is to read/write Microsoft installer files (.msi)

File: MissionAlertBot.py [hunk 2]
```diff
@@ -515,98 +524,65 @@ def _delete_all_from_database(database):
     carriers_conn.commit()
 
 
-# function to search for a carrier by longname
-def find_carrier_from_long_name(find_long_name):
+# function to search for a carrier
+def find_carrier(searchterm, searchfield):
     """
-    Finds any carriers matching a long name
+    Finds any carriers matching a searchterm
```
Comment (gcruicks) [inline MissionAlertBot.py:530]: Finds any carriers matching a searchterm by searching on a specific column.

File: MissionAlertBot.py [hunk 3]
```diff
@@ -515,98 +524,65 @@ def _delete_all_from_database(database):
...
+    :param str searchfield: DB column to match against
     :returns: CarrierData object for the exact match
     :rtype: CarrierData
     """
     # TODO: This needs to check an exact not a `LIKE`
     carrier_db.execute(
-        f"SELECT * FROM carriers WHERE longname LIKE (?)",
-        (f'%{find_long_name}%',))
+        f"SELECT * FROM carriers WHERE {searchfield} LIKE (?)",
+        (f'%{searchterm}%',))
```
Comment (gcruicks) [inline MissionAlertBot.py:540]: `"SELECT * FROM carriers WHERE ? LIKE (?)", (f'{searchfield}', f'%{searchterm}%',))`

I think that is the right syntax. Basically ensures the values get escaped correctly

File: MissionAlertBot.py [hunk 4]
```diff
@@ -515,98 +524,65 @@ def _delete_all_from_database(database):
...
-              f"shortname {carrier_data.carrier_short_name} with channel #{carrier_data.discord_channel} called from "
-              f"find_carrier_from_long_name.")
-
-    return carriers
 
-
-def find_carrier_with_owner_id(ownerid):
+# used to find carriers if we expect multiple results for a search term
+# TODO: make every carrier longname search prompt with multiple results and use this function
+def find_carriers_mult(searchterm, searchfield):
```
Comment (gcruicks) [inline MissionAlertBot.py:550]: This and `find_carriers` could be rolled into one function. Just return always a list of carrier objects, and iterate accordingly. They both call the same thing, just one is `fetchone` while the other `fetchall`. Just always fetch all, and you have either a single element list, or a multi element.

You could even take a `table` as a parameter and just have one function handle all the searches (missions, carriers, community carriers etc)

File: MissionAlertBot.py [hunk 5]
```diff
@@ -2356,7 +2311,7 @@ def check(message):
 @commands.has_any_role('Certified Carrier', 'Trainee')
 async def carrier_image(ctx, lookname):
     print(f"{ctx.author} called m.carrier_image for {lookname}")
-    carrier_data = find_carrier_from_long_name(lookname)
+    carrier_data = find_carrier(lookname, "longname")
```
Comment (gcruicks) [inline MissionAlertBot.py:2314]: We might want to consider an enumeration of fields,

class CarrierDbFields(enum):
   longname = 'longname',
  shortname = 'short name',
  etc ...

Then use `find_carrier(look name, CarrierDbFields.longname)`

Not critical now, but might be worth it.

File: MissionAlertBot.py [hunk 6]
```diff
@@ -3104,33 +3055,49 @@ def check(message):
...
+    #
+    # first see if there's a channel matching the given string as-is
     new_channel = discord.utils.get(ctx.guild.channels, name=stripped_channel_name)
-
+    
     if new_channel:
+        if new_channel.name.endswith('_cc'):
+            print(f"Channel detected matching user string, ends with _cc: {new_channel.name}")
+            # if the user used _cc in their channel name this will be true and we can proceed
+            reuse_channel = True if new_channel else False
```
Comment (gcruicks) [inline MissionAlertBot.py:3083]: this case is always going to be `True` here You gated on line 3079 `if newchannel` already

**Review by gcruicks — CHANGES_REQUESTED** [review]

File: MissionAlertBot.py (hunk 1 above)
Comment (cimspin) [reply MissionAlertBot.py:10]: Weird, I don't know why that's there. I definitely didn't put it there. Maybe Visual Studio Code did it automatically?

File: MissionAlertBot.py (hunk 4 above)
Comment (cimspin) [reply MissionAlertBot.py:550]: Plan is to merge them all one day but not today

File: MissionAlertBot.py (hunk 5 above)
Comment (cimspin) [reply MissionAlertBot.py:2314]: Implemented a probably less than ideal version of this, have to call using CarrierDbFields.longname.name etc.

File: MissionAlertBot.py (hunk 3 above)
Comment (cimspin) [reply MissionAlertBot.py:540]: Couldn't get this to work using any variation of this syntax, no idea why. More learning is needed

**Pull request merged**